### DISCUSSION TITLE
Add nodes for Blackboard operations

### DISCUSSION THREAD

**Discussion by Wichamir** [bitbrain/beehave#125] [Ideas]
When working with behavior trees I have to quite often check if blackboard has some key, or set it to some predefined value (like 0 or INF) or compare some values. It would be nice if there was a bunch of predefined nodes for blackboard operations. My idea is to add 4 nodes:

- `BlackboardSetAction`, which would take a `String` representing a key and and a `String` representing an `Expression`, which would be parsed and saved into said key
- `BlackboardEraseAction`, which would erase the specified key from the blackboard
- `BlackboardHasCondition`, which would simply return `SUCCESS` if blackboard has a specified key, otherwise FAILURE
- `BlackboardCompareCondition`, which would take two Strings representing `Expression`s and an enum representing a comparison operator; when ticked it would evaluate both expressions and compare their values returning either `SUCCESS` or `FAILURE`

Thoughts?

**Comment by bitbrain** [comment]
Great suggestion. I really like this idea and these nodes could be very useful for a lot of folks.

One challenge here is to define, what qualifies as "generic enough" to be added as a node. I fear that once we start adding predefined actions and conditions, more and more people want other custom nodes. Eventually, we could end up with dozens or even hundreds of nodes that are very helpful each but only to some people.

Ideally, this addon should be kept as generic and simple as possible. Agreeing on a set of rules first might be the difficult part here.

**Reply by Wichamir** [reply]
Well I think we can define "generic" as "not related to any specific type of game, but to the functionality of Beehave itself, while being commonly useful". In the case of blackboard operations they're strictly related to the Blackboard class and not some external symbol and they seem to be quite useful no matter what type of game you're making.
